text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+          backwards compatible for some functions (except for on a trait's
+          associated type). An existing function with a `Sized` or `ValueSized`
+          bound could be relaxed to the new trait without breaking existing code.
+          Existing functions with a `Pointee` bound could not.
+      - After `Pointee`
+        - i.e. `Sized: ValueSized: Pointee: NewSized`
+        - For the same reasons as the traits proposed in this RFC, adding bounds
+          on a new trait after `Pointee` is backwards compatible (except for on a
+          trait's associated type). Any existing function will have stricter
+          bounds than the new trait and so could be relaxed to the new trait
+          without breaking existing code.
+
+## `Sized` bounds
+[sized-bounds]: #sized-bounds
+
+`?Sized` would be made syntactic sugar for a `const ValueSized` bound. A
+`const ValueSized` bound is equivalent to a `?Sized` bound as all values in Rust
+today whose types do not implement `Sized` are valid arguments to
+[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
+computed given a value and knowledge of the type and target platform, and
+therefore will implement `const ValueSized`. As there are currently no
+extern types or other types which would not implement `const ValueSized`,
+every type in Rust today which would satisfy a `?Sized` bound would satisfy
+a `const ValueSized` bound.
+
+**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
+a `const ValueSized` bound. As the `?Trait` syntax is currently accepted for any trait
+but ignored for every trait except `Sized`, `?ValueSized` and `?Pointee` bounds would
+be ignored. In the next edition, any uses of `?Sized` syntax will be rewritten to
+a `const ValueSized` bound. Any other uses of the `?Trait` syntax will be removed as
+part of the migration and the `?Trait` syntax will be prohibited.
+
+A default implicit bound of `const Sized` is added by the compiler to every type
+parameter `T` that does not have an explicit `~const Sized`, `Sized`, `?Sized`,
+`const ValueSized`, `~const ValueSized`, `ValueSized` or `Pointee` bound. It is
+backwards compatible to change the current implicit `Sized` bound to an `const Sized`
+bound as every type which exists currently will implement `const Sized`.
+
+**Edition change:** In the current edition, all existing `Sized` bounds will be
+syntatic sugar for `const Sized` (both explicitly written and implicit). In the next
+edition, existing `Sized` bounds will be rewritten to `const Sized` bounds and new
+bare `Sized` bounds will be non-const. `Sized` supertraits (such as that on `Clone`
+or `Default`) will not be sugar for `const Sized`, these will remain bare `Sized`.
+If traits with a `Sized` supertrait are later made const, then their supertrait
+would be made `~const Sized`.
+
+An implicit `const ValueSized` bound is added to the `Self` type of traits. Like
```

Comment:
I've also added this as an example in the RFC.

Reply:
I disagree, the difference between having no default supertrait and having one is massive. Not having a default supertrait means that traits naturally apply to all types and library authors don't have to remember (or be asked) to relax that bound where possible (which would be a breaking change).

Not being able to implement any std traits on these types feels like a big problem to me, it would make them a nightmare to use without Debug, Clone, PartialEq, etc. Admittedly, this is probably less of a problem for the vector types as it's less clear (to me at least) what they could implement given they might contain uninitialised bytes. I wonder if it would be possible to relax the bounds on existing traits by introducing explicit bounds on traits with default implementations.

Reply:
> I disagree, the difference between having no default supertrait and having one is massive.

I wasn't intending to argue that this limitation wasn't significant, just that I didn't think it was hard to teach or remember.

Reply:
I've pushed another change related to this, I think the implicit bound on `Self` is necessary but it should be possible to relax it backwards-compatibily.